modules/compiler/vecz/source/pass.cpp

```diff
@@ -262,4 +271,96 @@ std::optional<VeczPassOptions> getReqdSubgroupSizeOpts(Function &F) {
   return std::nullopt;
 }
 
+std::optional<VeczPassOptions> getAutoSubgroupSizeOpts(
+    Function &F, ModuleAnalysisManager &AM) {
+  // If there's a required sub-group size, we must return a vectorization
+  // factor that gets us there.
+  if (auto opts = getReqdSubgroupSizeOpts(F)) {
+    return opts;
+  }
+
+  auto &M = *F.getParent();
+  const auto &GSGI = AM.getResult<compiler::utils::SubgroupAnalysis>(M);
+
+  // If the function doesn't use sub-groups (from the user's perspective) then
+  // we don't need to adhere to a specific sub-group size.
+  if (!GSGI.usesSubgroups(F)) {
+    return std::nullopt;
+  }
+
+  // Use the device's sub-group sizes to determine which to vectorize to.
+  auto &DI = AM.getResult<compiler::utils::DeviceInfoAnalysis>(M);
+
+  // We don't force devices to support any sub-group sizes.
+  if (DI.reqd_sub_group_sizes.empty()) {
+    return std::nullopt;
+  }
+
+  vecz::VeczPassOptions vecz_opts;
+  vecz_opts.vec_dim_idx = 0;
+  // Disable auto - we want a specific width
+  vecz_opts.vecz_auto = false;
+  // Enable some default choices
+  vecz_opts.choices.enable(vecz::VectorizationChoices::eDivisionExceptions);
+
+  // Now try and choose the best width.
+  std::optional<unsigned> best_width;
+  auto const mux_sub_group_size = compiler::utils::getMuxSubgroupSize(F);
+
+  auto can_produce_legal_width = [&mux_sub_group_size](unsigned size) {
+    // We only support vectorization widths where there's a clean multiple, and
+    // we can vectorize *up* to the desired size - we can't shrink the
+    // sub-group size by vectorizing.
+    return size >= mux_sub_group_size && size % mux_sub_group_size == 0;
+  };
+
+  for (auto size : DI.reqd_sub_group_sizes) {
+    if (!can_produce_legal_width(size)) {
```

Comment:
if i understand this correctly it will only replace the first one if it is a:
*non-scalar
* required WGS related
* power of 2

Does this mean we have a preferred ordering in terms of priority for  DI.reqd_sub_group_sizes

Reply:
Aye that's right. Well, plus "not 1". I didn't want to encode device-specific logic at this stage.

I think I documented that it's preferred ordering, but documenting it in the mux side of things seems wrong to me as I don't think we're ready to make it part of the mux "spec". This is more of a compiler implementation detail.